src/GraphQL/Output.hs

```diff
@@ -0,0 +1,76 @@
+-- | GraphQL output.
+--
+-- How we encode GraphQL responses.
+module GraphQL.Output
+  ( Response(..)
+  , Errors
+  , Error(..)  -- XXX: Maybe export helper functions rather than constructors.
+  ) where
+
+import Protolude hiding (Location, Map)
+
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.Map as Map
+import GraphQL.Value (Map, ToValue(..), Value(ValueNull))
+
+-- | GraphQL response.
+--
+-- A GraphQL response must:
+--
+--   * be a map
+--   * have a "data" key iff the operation executed
+--   * have an "errors" key iff the operation encountered errors
+--   * not include "data" if operation failed before execution (e.g. syntax errors,
+--     validation errors, missing info)
+--   * not have keys other than "data", "errors", and "extensions"
+--
+-- Other interesting things:
+--
+--   * Doesn't have to be JSON, but does have to have maps, strings, lists,
```

Comment:
interestingly it's almost like a protobuf - and I'm sure we could encode it as that (but for internal RPC sth like a `Binary` encoding might be faster)